dcrtimed/backend/filesystem/dump.go

```diff
@@ -370,3 +376,479 @@ func (fs *FileSystem) Restore(f *os.File, verbose bool, location string) error {
+			return fmt.Errorf("   *** ERROR hash not found in "+
+				"timestamp : %v %v\n", filepath.Join(fs.root,
+				ts2dirname(value)), key)
+		}
+
+		// Check to see if this timestamp exists in the empties map.
+		// This is bad because if it is we have an empty timestamp
+		// directory while the global directory thinks there should be
+		// a record.
+		if _, ok := empties[value]; ok {
+			// There really is no good recovery from this. A
+			// potential solution would be to create a record but
+			// it may be inaccurate.
+			//
+			// At this point we treat this as fatal because there
+			// is no need to write code for theoretical issues that
+			// should not have happned.
+			return fmt.Errorf("   *** ERROR global pointer to "+
+				"empty timestamp: %v (%v)", ts2dirname(value),
+				value)
+		}
+	}
+	if i.Error() != nil {
+		return i.Error()
+	}
+
+	// Range over empties and try to delete them.
+	for k := range empties {
+		// It is safe to delete the timestamp directory. The reason
+		// these directories exist is because leveldb has a bug that it
+		// always creates the container directory despite being told
+		// not to. This bug was not worked around in early dcrtime
+		// deployments but has since been fixed.
+		fmt.Printf("   *** FIXING removing empty timestamp: %v (%v)\n",
+			ts2dirname(k), k)
+		path := filepath.Join(fs.root, ts2dirname(k))
+		err := os.RemoveAll(path)
+		if err != nil {
+			return fmt.Errorf("   *** ERROR RemoveAll: %v", err)
+		}
+		delete(empties, k)
+	}
+	// Make sure we don't ave any empties left over
+	if len(empties) != 0 {
+		// Shouldn'y happen
+		return fmt.Errorf("   *** ERROR empties not pruned: %v ",
+			spew.Sdump(empties))
+	}
+
+	return nil
+}
+
+// fsckDup checks for duplicate digests in the global dups map.
+func (fs *FileSystem) fsckDup(options *backend.FsckOptions, ts int64, dups map[string]int64) error {
+	db, err := fs.openRead(ts)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	i := db.NewIterator(nil, nil)
+	defer i.Release()
+	for i.Next() {
+		key := i.Key()
+		if string(key) == flushedKey {
+			// Skip FlushRecord
+			continue
+		}
+		k := hex.EncodeToString(key)
+		if v, ok := dups[k]; ok {
+			// This is the result of a bug that was caused when the
+			// server was shutdown over a flush window and the user
+			// submitted the same hash again.
+			// The bug that caused this has since been fixed.
+			//
+			// The only thing we can do is assert (which should
+			// have been caught!) that there is no entry in the
+			// global db and delete the duplicate entry. This is
+			// safe because the caller has not caused a flush yet
+			// and therefore has no record YET of which TX it lives
+			// in.
+			fmt.Printf("    *** ERROR duplicate key: %v %v %v\n",
+				ts, v, k)
+
+			// 1. Verify against global db
+			ok, err := fs.db.Has(key, nil)
+			if err != nil {
+				return fmt.Errorf("   *** ERROR duplicate "+
+					"key: has %v", err)
+			}
+			if ok {
```

Comment:
Won't `ok` always be true since we already know that the digest is a duplicate?  

Don't we need to make sure that we're deleting the correct duplicate by retrieving the timestamp from the global database and checking if it matches `ts`.  If it doesn't match then we delete the digest from the current timestamp db.

Reply:
Actually the other way around. The test is in there to make sure we don't catch false-positives.

So the dup list ONLY is filled when there is no flush record AND when there are multiple entries for this hash. This means that since it has never been flushed it does not matter which one is deleted and one could argue that the first one IS the correct one since the follow up ones were newer submissions due to the bug described in the comments. I think this is working correctly.

Reply:
> So the dup list ONLY is filled when there is no flush

Maybe I'm missing something, but I'm not seeing this in the code.

Reply:
The dups map is only filled when there are NO flush records. Look at the code around line 750. There is a continue if there is a flush record.

Reply:
Sorry, that is incorrect. Let me rephrase.

This code iterates over all timestamps and adds all digests to a map. As it is building the map it looks for dups (which is an invalid condition). When it finds a dup it ensures it doesn't exist in the global database (meaning there is no flush record!). I added code that verifies if it does exist in the database that it errors out. Also added code to print all hashes and stuff during phase 3 (when verbose and printhashes) so that these claims can be cross verified.